### PR TITLE
Improve tests for catch-all + defaults/nulls

### PR DESCRIPTION
**Pull request by rynowak**

### CODE REVIEW

**Comment by rynowak** [comment]
I noticed some squishiness + missing tests while investigating something in this area

File: test/Microsoft.AspNetCore.Routing.Tests/Tree/TreeRouterTest.cs [hunk 1]
```diff
@@ -1581,9 +1672,19 @@ private static TreeRouteMatchingEntry CreateMatchingEntry(IRouter router, string
...
+            var defaults = new RouteValueDictionary();
+            foreach (var parameter in parsedRouteTemplate.Parameters)
+            {
+                if (parameter.DefaultValue != null)
+                {
+                    defaults.Add(parameter.Name, parameter.DefaultValue);
+                }
+            }
+
+            defaults["test_route_group"] = routeGroup;
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Routing.Tests/Tree/TreeRouterTest.cs]: The MVC code makes the same mistake here :cry: see aspnet/Mvc#4447

File: test/Microsoft.AspNetCore.Routing.Tests/Template/TemplateMatcherTests.cs [hunk 2]
```diff
@@ -751,12 +751,30 @@ public void TryMatch_RouteWithCatchAllClauseCapturesEmptyContent()
...
+
+            // Assert
+            Assert.True(match);
+            Assert.Equal<int>(2, values.Count);
+            Assert.Equal("v1", values["p1"]);
+            Assert.Equal("hello", values["p2"]);
+        }
+
+        [Fact]
+        public void TryMatch_RouteWithCatchAll_UsesDefaultValueForEmptyContent()
```
Comment (javiercn) [inline test/Microsoft.AspNetCore.Routing.Tests/Template/TemplateMatcherTests.cs]: Is the name of this test accurate? You aren't checking the default value, you are checking the one that you override.

Comment (javiercn) [inline test/Microsoft.AspNetCore.Routing.Tests/Template/TemplateMatcherTests.cs]: Nevermind, I'm wrong

File: test/Microsoft.AspNetCore.Routing.Tests/Tree/TreeRouterTest.cs [hunk 3]
```diff
@@ -153,6 +153,97 @@ public class TreeRouterTest
...
+
+            // Assert
+            Assert.NotNull(context.Handler);
+            Assert.Null(context.RouteData.Values["path"]);
+        }
+
+        [Theory]
+        [InlineData("a/{*path}", "/a")]
+        [InlineData("a/{*path}", "/a/")]
+        public async Task TreeRouter_RouteAsync_MatchesCatchAll_NullValue_DoesNotTrounceExistingValue(
```
Comment (javiercn) [inline test/Microsoft.AspNetCore.Routing.Tests/Tree/TreeRouterTest.cs]: nit: Isn't there a better word than trounce? Like replace, win, or something more common in english?

**Comment by javiercn** [comment]
Looks good to me

**Comment by javiercn** [comment]
:shipit:

File: test/Microsoft.AspNetCore.Routing.Tests/Template/TemplateMatcherTests.cs [hunk 4]
```diff
@@ -751,12 +751,30 @@ public void TryMatch_RouteWithCatchAllClauseCapturesEmptyContent()
...
+            var matcher = CreateMatcher("{p1}/{*p2}");
+
+            var values = new RouteValueDictionary(new { p2 = "hello" });
+
+            // Act
+            var match = matcher.TryMatch("/v1", values);
+
+            // Assert
+            Assert.True(match);
+            Assert.Equal<int>(2, values.Count);
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Routing.Tests/Template/TemplateMatcherTests.cs]: Why do you have to do `<int>`?

File: test/Microsoft.AspNetCore.Routing.Tests/Tree/TreeRouterTest.cs (hunk 3 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Routing.Tests/Tree/TreeRouterTest.cs]: Changed to say "replace"

File: test/Microsoft.AspNetCore.Routing.Tests/Template/TemplateMatcherTests.cs (hunk 4 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Routing.Tests/Template/TemplateMatcherTests.cs]: This was a copy-paste

**Pull request closed without merging**